src/EditorFeatures/Core/Implementation/AddImports/AddImportsPasteCommandHandler.cs

```diff
@@ -0,0 +1,120 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddMissingImports;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Editor.Shared.Options;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.AddImports
+{
+    [Export]
+    [Export(typeof(ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.AddImportsPaste)]
+    // Order is important here, this command needs to execute before PasteTracking
+    // since it may modify the pasted span. Paste tracking dismisses if 
+    // the span is modified. It doesn't need to be before FormatDocument, but
+    // this helps the order of execution be more constant in case there 
+    // are problems that arise. This command will always execute the next
+    // command before doing operations.
+    [Order(After = PredefinedCommandHandlerNames.PasteTrackingPaste)]
+    [Order(Before = PredefinedCommandHandlerNames.FormatDocument)]
+    internal class AddImportsPasteCommandHandler : IChainedCommandHandler<PasteCommandArgs>
+    {
+        public string DisplayName => EditorFeaturesResources.Add_Missing_Imports_On_Paste;
+
+        private readonly IThreadingContext _threadingContext;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public AddImportsPasteCommandHandler(IThreadingContext threadingContext)
+            => _threadingContext = threadingContext;
+
+        public void ExecuteCommand(PasteCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Capture the pre-paste caret position
+            var caretPosition = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caretPosition.HasValue)
+            {
+                return;
+            }
+
+            nextCommandHandler();
+
+            if (!args.SubjectBuffer.CanApplyChangeDocumentToWorkspace())
+            {
+                return;
+            }
+
+            // Don't perform work if we're inside the interactive window
+            if (args.TextView.IsBufferInInteractiveWindow(args.SubjectBuffer))
+            {
+                return;
+            }
```

Comment:
this means we don't support this for razor.  is that ok?

Reply:
I think it's fine for now. I'll follow up with razor folks to see if it is something they want enabled.